### ISSUE TITLE
BLD: build failure with new setuptools 73.0.0

### ISSUE THREAD

**Issue by tylerjereddy** [scipy/scipy#21416]
I don't think it is particularly surprising when this happens, but it did just bite me while working on the release/backports in gh-21362, and I confirmed that I can reproduce on `main` with `setuptools` `73.0.0`, released an hour ago, via its interaction with `pythran` `0.16.0`, `0.15.0`, or `0.16.1` on MacOS ARM (didn't check other platforms).

<details>

```
[560/1283] Generating 'scipy/stats/_stats_pythran.cpython-312-darwin.so.p/_stats_pythran.cpp'
FAILED: scipy/stats/_stats_pythran.cpython-312-darwin.so.p/_stats_pythran.cpp 
/Users/treddy/python_venvs/py_312_scipy_dev2/bin/pythran -E ../scipy/stats/_stats_pythran.py -o scipy/stats/_stats_pythran.cpython-312-darwin.so.p/_stats_pythran.cpp
Traceback (most recent call last):
  File "/Users/treddy/python_venvs/py_312_scipy_dev2/bin/pythran", line 8, in <module>
    sys.exit(run())
             ^^^^^
  File "/Users/treddy/python_venvs/py_312_scipy_dev2/lib/python3.12/site-packages/pythran/run.py", line 190, in run
    pythran.compile_pythranfile(args.input_file,
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/treddy/python_venvs/py_312_scipy_dev2/lib/python3.12/site-packages/pythran/__init__.py", line 127, in __getattr__
    import pythran.toolchain
  File "/Users/treddy/python_venvs/py_312_scipy_dev2/lib/python3.12/site-packages/pythran/toolchain.py", line 11, in <module>
    from pythran.dist import PythranExtension, PythranBuildExt
  File "/Users/treddy/python_venvs/py_312_scipy_dev2/lib/python3.12/site-packages/pythran/dist.py", line 142, in <module>
    class PythranBuildExt(PythranBuildExtMixIn, LegacyBuildExt, metaclass=PythranBuildExtMeta):
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
[575/1283] Compiling C++ object scipy/sparse/sparsetools/_sparsetools.cpython-312-darwin.so.p/bsr.cxx.o
```

</details>

`python -m pip install "setuptools<73.0.0"` in an affected environment suffices to fix the issue. I suspect this is an "upstream" matter, but perhaps worth noting at least. cc @serge-sans-paille just in case maybe, though perhaps `setuptools` will yank if it is a problem on their end.

**Comment by tylerjereddy** [comment]
This hack does seem to help isolated builds deal with the transitive dep issue locally, but is obviously very undesirable:

```diff
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -28,6 +28,7 @@ requires = [
     # The upper bound on pythran is pre-emptive only; 0.16.1
     # is released/working at time of writing.
     "pythran>=0.14.0,<0.17.0",
+    "setuptools<73.0.0", # hack to work around gh-21416
```

So, hopefully gets resolved/yanked upstream.

**Comment by rgommers** [comment]
xref https://github.com/serge-sans-paille/pythran/issues/2228#issuecomment-2299000192 for context

**Comment by agriyakhetarpal** [comment]
I just bisected, and the first bad commit in `setuptools` seems to be [f9398b1f](https://github.com/pypa/setuptools/commit/f9398b1f2bd6cce0aa4efa01d8b69da021205151). I'll have a look and try to write a bug report for this there.

_cross-posted from https://github.com/serge-sans-paille/pythran/issues/2228_

**Comment by agriyakhetarpal** [comment]
Never mind, that doesn't seem to be the correct bad commit – bisecting again

**Comment by agriyakhetarpal** [comment]
Upstream issue: https://github.com/pypa/setuptools/issues/4579 (as linked above).

**Comment by agriyakhetarpal** [comment]
New tag: https://github.com/pypa/setuptools/releases/tag/v73.0.1

**Comment by agriyakhetarpal** [comment]
This can now be closed since https://pypi.org/project/setuptools/73.0.1/ is now up.

**Comment by tylerjereddy** [comment]
@agriyakhetarpal Thanks for the hard work bisecting and coordinating upstream with `setuptools` for us.

I see the issue fixed locally as well, with the newly-released `setuptools` `73.0.1`.
